usng.js

```diff
@@ -382,6 +462,166 @@
             return result;
         },
 
+        serializeUPS(upsCoordinates = {}) {
+            //verify that the passed in coords is a valid ups object
+            try { this.UPStoLL(upsCoordinates) } 
+            catch (err) { throw err }
+
+            const {northPole, easting, northing} = upsCoordinates
+            const zoneLetter = northPole ? (easting < 2000000 ? 'Y' : 'Z') 
+                : (easting < 2000000 ? 'A' : 'B')
+            return `${zoneLetter} ${easting}mE ${northing}mN` 
+        },
+        
+        deserializeUPS(str) {
+            let zoneLetter, easting, northing
+
+            const processInvalidUPS = () => {
+                throw new Error( `Invalid UPS String: ${str}`)
+            }
+
+            const processValidUPS = () => {
+                return {
+                    northPole: (["Y", "Z"].includes(zoneLetter.toUpperCase())),
+                    easting: Number(easting),
+                    northing: Number(northing)
+                }
+            }
+
+            const regExp = /^\s*([abyzABYZ])\s+(\d+)[mM][eE]\s+(\d+)[mM][nN]\s*$/
+            try {
+                [, zoneLetter, easting, northing] = regExp.exec(str)
+                this.UPStoLL({
+                    northPole: true, 
+                    easting: Number(easting),
+                    northing: Number(northing)
+                })
+            } catch(err) {
+                processInvalidUPS()
+            }
+            
+            return processValidUPS()
+        },
+
+        /***************** convert latitude, longitude to UPS  *******************
+         Uses "northPole" boolean instead of returning a negative Northing value
+
+         Input latitude is expected to be [-90, 90].
+         Input longitude is expected to be [-180, 180].
+
+         This returns an object describing the UPS coordinates:
+         { northing: number, easting: number, northPole: bool }
+
+         On invalid input throws an error.
+        ***************************************************************************/
+        LLtoUPS: function(lat, lon) {
+            const isLatLonCoordinatesPairInvalid = () => {
+                return !(typeof lat === "number"
+                    && typeof lon === "number"
+                    && lat >= -90 && lat <= 90
+                    && lon >= -180 && lon <= 180)
+            }
+            if (isLatLonCoordinatesPairInvalid()) {
+                throw new Error('Invalid Lat/Lon coordinates: ' + lat + ', ' + lon)
+            }
+            const northPole = lat > 0
+            const tau = Math.tan(Math.abs(lat) * this.DEG_2_RAD)
+            const taup = taupf.call(this, tau)
+            const rhoStep1 = Math.hypot(1, taup) + Math.abs(taup)
+            const rhoStep2 = taup >= 0
```

Comment:
✏️ These nested ternaries are incredibly difficult to read. Could create functions or use explicit if/else semantics instead.

Reply:
I would argue it depends on the formatting, for example if Python or Haskell like style is used they are easier to read than explicit `if-else` trees (while trying to keep track of mutability).
As a compromise maybe this:
```javascript
const rhoStep2 =  Math.abs(lat) !== 90 ? 1/rhoStep1 : 0
const rhoStep3 = taup >= 0 ? rhoStep2 : rhoStep1
const rho = rhoStep3 * this.rhoAdjusterValue
```

Reply:
That compromise is 💯 with me.